usr.sbin/mrinfo/mrinfo.8
.\"	$NetBSD: mrinfo.8,v 1.8 2003/03/31 01:34:57 perry Exp $
.\"
.\" Written Wed Mar 24 1993 by Van Jacobson (adapted from the
.\" multicast mapper written by Pavel Curtis).
.\"
.\" The lawyers insist we include the following UC copyright notice.
.\" The mapper from which this is derived contained a Xerox copyright
.\" notice which follows the UC one.  Try not to get depressed noting
.\" that the legal gibberish is larger than the program.
.\"
.\" Copyright (c) 1993 Regents of the University of California.
.\" All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. All advertising materials mentioning features or use of this software
.\"    must display the following acknowledgement:
.\"	This product includes software developed by the Computer Systems
.\"	Engineering Group at Lawrence Berkeley Laboratory.
.\" 4. Neither the name of the University nor of the Laboratory may be used
.\"    to endorse or promote products derived from this software without
.\"    specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\" ---------------------------------
.\" Copyright (c) 1992, 2001 Xerox Corporation.  All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions are met:
.\"
.\" Redistributions of source code must retain the above copyright notice,
.\" this list of conditions and the following disclaimer.
.\"
.\" Redistributions in binary form must reproduce the above copyright notice,
.\" this list of conditions and the following disclaimer in the documentation
.\" and/or other materials provided with the distribution.
.\"
.\" Neither name of the Xerox, PARC, nor the names of its contributors may be
.\" used to endorse or promote products derived from this software
.\" without specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
.\" ``AS IS'' AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
.\" TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
.\" PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE XEROX CORPORATION OR
.\" CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
.\" EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
.\" PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS;
.\" OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
.\" WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR
.\" OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE,
.\" EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
.\"
.TH MRINFO 8
.UC 5
.SH NAME
mrinfo \- Displays configuration info from a multicast router
.SH SYNOPSIS
.B /usr/sbin/mrinfo
[
.B \-d
.I debug_level
] [
.B \-r
.I retry_count
] [
.B \-t
.I timeout_count
]
.B multicast_router

.SH DESCRIPTION
.I mrinfo
attempts to display the configuration information from the multicast router
.I multicast_router.
.PP
.I mrinfo
uses the ASK_NEIGHBORS IGMP message to the specified multicast router. If this
multicast router responds, the version number and a list of their neighboring
multicast router addresses is part of that response. If the responding router
has a recent multicast version number, then
.I mrinfo
requests additional information such as metrics, thresholds, and flags from the
multicast router. Once the specified multicast router responds, the
configuration is displayed to the standard output.
.br
.ne 5
.SH INVOCATION
.PP
"\-d" option sets the debug level. When the debug level is greater than the
default value of 0, addition debugging messages are printed. Regardless of
the debug level, an error condition, will always write an error message and will
cause
.I mrinfo
to terminate.
Non-zero debug levels have the following effects:
.IP "level 1"
packet warnings are printed to stderr.
.IP "level 2"
all level 1 messages plus notifications down networks are printed to stderr.
.IP "level 3"
all level 2 messages plus notifications of all packet
timeouts are printed to stderr.
.PP
"\-r retry_count" sets the neighbor query retry limit. Default is 3 retry.
.PP
"\-t timeout_count" sets the number of seconds to wait for a neighbor query
reply. Default timeout is 4 seconds.
.PP
.SH SAMPLE OUTPUT
.nf
.I mrinfo mbone.phony.dom.net
127.148.176.10 (mbone.phony.dom.net) [version 3.3]:
 127.148.176.10 -\*[Gt] 0.0.0.0 (?) [1/1/querier]
 127.148.176.10 -\*[Gt] 127.0.8.4 (mbone2.phony.dom.net) [1/45/tunnel]
 127.148.176.10 -\*[Gt] 105.1.41.9 (momoney.com) [1/32/tunnel/down]
 127.148.176.10 -\*[Gt] 143.192.152.119 (mbone.dipu.edu) [1/32/tunnel]
.fi
.PP
For each neighbor of the queried multicast router, the IP of the queried router
is displayed, followed by the IP and name of the neighbor. In square brackets
the metric (cost of connection), the threshold (multicast ttl) is displayed. If
the queried multicast router has a newer version number, the type (tunnel,
srcrt) and status (disabled, down) of the connection is displayed.
.PP
.SH IMPORTANT NOTE
.I mrinfo
must be run as root.
.PP
.SH SEE ALSO
.BR mrouted (8) ,
.BR map-mbone (8) ,
.BR mtrace (8)
.PP
.SH AUTHOR
Van Jacobson


doc/CHANGES
# LIST OF CHANGES FROM LAST RELEASE:			<$Revision: 1.1142.2.10 $>
#
#
# [Note: This file does not mention every change made to the NetBSD source tree.
# For an exhaustive list, please subscribe to the `source-changes' mailing list,
# or see the mailing list archives.  For more information, send a message
# containing just the word `help' to `majordomo@NetBSD.org'.]
#
# ----------------------------------------------------------------------------
# PLEASE USE THE FOLLOWING STYLE WHEN ADDING ITEMS:
#
#	theme: Content ...
#		more content ...
#		more content. [developer DATE-FORMAT]
#
# That's it, one tab to start an item, and two tabs to indent the content.
# Also please start the content with a capital and end it with a dot.
#
# This is very important for the www scripts, otherwise it won't be able to
# parse the file and the HTML output will be mangled.
#
# See htutils/changes/changes2html script for more details.
#
# ----------------------------------------------------------------------------


Changes from NetBSD 4.0 to NetBSD 5.0:
	evbppc: Add support for Xilinx Virtex II-Pro/4-FX. [freza 20061202]
	mac68k: Add support for four more NuBus video cards, including the
		VillageTronic MacPicasso 340. [hauke 20061206]
	puc(4): Add support for Perle Systems 4 and 8 ports modem boards.
		[manu 20071209]
	arm: Add support for NPE Ethernet on IXP425. [scw 20061210]
	arm: Add support for IXP425 watchdog timer. [scw 20061210]
	libutil: Add flags_to_string(3) and string_to_flags(3),
		moved from ls(1). [christos 20061214]
	arm: Add support for i2c, i2s, pcmcia, and ohci of PXA2x0.
		Ported from OpenBSD as part of zaurus port by NONAKA Kimihiro.
		[peter 20061217]
	zaurus: Add support for Sharp C3x00 PDA. [ober 20061217]
	kernel: Merge yamt-splraseipl branch, which provides splraiseipl(9)
		and makeiplcookie(9). [yamt 20061221]
	sgimips: Add support for Set Engineering's GIO Fast Ethernet board.
		[rumble 20061223]
	sgimips: Add support for the SGI Light/Entry/Starter LG1/LG2
		framebuffers found in Indigo systems. [rumble 20061226]
	nfs: Remove NQNFS support from kernel. [yamt 20061227]
	kernel, srtconfig(1): First import of the source-address-based routing
		pseudo-device.  [mouse 20061229]
	sgimips: Add i8254 timecounter(9) support for use on IP12.
		[rumble 20061229]
	mount_psshfs(8): Add puffs sshfs. [pooka 20061229]
	njata(4): Add support for Workbit CF32A CF adapter,
		aka BUFFALO RCF-CBA2.  [ichiro 20061231]
	amd64, i386: Report and work around some of the known errata for
		Athlon 64 and Opteron processors. [ad 20070101]
	i386: Add fatboot, sub-512 byte bootloader that can be written into
		the first sector of a FAT16 filesystem and will load
		the secondary bootloader from it. [dsl 20070101]
	kernel: Make mount(2) and unmount(2) use kauth(9) for security
		policy.  One incompatible change is that an unprivileged
		mount without the MNT_NOSUID and MNT_NODEV flags
		will now fail, whereas previously those flags were
		automatically added.  [elad 20070102]
	udf: Implement read-only access to UDF 2.50 and UDF 2.60 discs
		that use metadata partition mappings. [reinoud 20070104]
	i386: Add support for Microsoft Xbox. [jmcneill 20070105]
	ntp: Import ntp 4.2.4 (20061228). Tracking of network interface
		address changes is now supported - no more ntpd
		restarts when interface addresses are re-configured
		(WLAN/DSL/ppp/...). [kardel 20070106]
	config(1): Introduce a new -L option, generates a lint configuration.
		[cube 20070108]
	umass(4): Add a quirk for the Sony Cybershot USB camera.
		From Mihai Chelaru. [christos 20070109]
	kernel: Global 'securelevel' variable removed; it is now internal to
		the secmodel. [elad 20070110]
	gcc(1): Updated to version 4.1.2-20070110. [mrg 20070111]
	zaurus: Add MI todr(9) and timecounter(9) support. [ober 20070112]
	acpitools: Added from FreeBSD via fvdl. [christos 20070114]
	kernel: Add syscalls for POSIX Real-time library. [ad 20070116]
	adb(4): Add new and simplified MI ADB drivers. [macallan 20070117]
	macppc: Add a driver for the SGS 7433 mixer found in some G3 Macs.
		[macallan 20070117]
	viaenv(4): Add support for the VIA VT8231 hardware monitor and
		remove viapm(4) code. From OpenBSD. [xtraeme 20070120]
	scsictl(8): Added setspeed command, to set the highest speed that the
		optical drive should use for reading data,
		contributed by Kimura Fuyuki. [bouyer 20070122]
	msk(4): Pull support for some devices and many bug fixes from OpenBSD.
		[msaitoh 20070130]
	uplcom(4): Add support for Willcom WS002IN PHS device
		(Prolific Technology PL2303X). [msaitoh 20070131]
	zaurus: Updates to allow Zaurus screen to rotate 90 degrees to a usable
		state with the keyboard. From OpenBSD. [ober,peter 20070202]
	security(8): Add support for per-user /tmp. [elad 20070204]
	ugensa(4): Add support Novatel Merlin, used in VZ V620.
		From John Nielsen on freebsd-mobile. [gdt 20070205]
	i386, amd64: Add cmos(4), driver for x86 CMOS RAM. [dyoung 20070206]
	find(1): Add -delete option from FreeBSD. [elad 20070206]
	rum(4): Add support for MELCO WLI-U2-SG54HP and PLANEX GW-US54Mini2.
		[elad 20070206]
	kernel: Merge the newlock2 branch. Introduces a number of new
		kernel synchronization primitives, improves scalability
		on MP systems, and replaces the existing SA threads model
		with a 1:1 threading model. [ad 20070209]
	news68k: Implement generic softintr(9) support. [tsutsui 20070210]
	ubsa(4): Add support for CDMA modems sold by Eurotel (now O2).
		[salo 20070210]
	refuse: Added FUSE (file system in userspace) compatibility
		functionality, layered on top of libpuffs. [agc 20070211]
	i386: Added an opencrypto provider for VIA ACE (AES encryption
		instructions), ported from OpenBSD. [daniel 20070217]
	shark: Add generic softintr(9) support. [matt 20070218]
	kernel: Add code to support per-system call statistics. [dsl 20070218]
	sdiff(1): Replaced by OpenBSD sdiff(1) (public domain). [rmind 20070219]
	sgimips: Support the sq(4) interface on the Challenge S's IOPLUS

usr.sbin/mrinfo/mrinfo.c
/*	$NetBSD: mrinfo.c,v 1.26 2007/02/22 01:29:35 hubertf Exp $	*/

/*
 * This tool requests configuration info from a multicast router
 * and prints the reply (if any).  Invoke it as:
 *
 *	mrinfo router-name-or-address
 *
 * Written Wed Mar 24 1993 by Van Jacobson (adapted from the
 * multicast mapper written by Pavel Curtis).
 *
 * The lawyers insist we include the following UC copyright notice.
 * The mapper from which this is derived contained a Xerox copyright
 * notice which follows the UC one.  Try not to get depressed noting
 * that the legal gibberish is larger than the program.
 *
 * Copyright (c) 1993 Regents of the University of California.
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. All advertising materials mentioning features or use of this software
 *    must display the following acknowledgement:
 *	This product includes software developed by the Computer Systems
 *	Engineering Group at Lawrence Berkeley Laboratory.
 * 4. Neither the name of the University nor of the Laboratory may be used
 *    to endorse or promote products derived from this software without
 *    specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 * ---------------------------------
 * Copyright (c) 1992, 2001 Xerox Corporation.  All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without modification,
 * are permitted provided that the following conditions are met:
 *
 * Redistributions of source code must retain the above copyright notice,
 * this list of conditions and the following disclaimer.
 *
 * Redistributions in binary form must reproduce the above copyright notice,
 * this list of conditions and the following disclaimer in the documentation
 * and/or other materials provided with the distribution.

 * Neither name of the Xerox, PARC, nor the names of its contributors may be used
 * to endorse or promote products derived from this software
 * without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS ``AS IS''
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
 * THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
 * PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE XEROX CORPORATION OR CONTRIBUTORS
 * BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
 * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS;
 * OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
 * WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR
 * OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE,
 * EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

#include <sys/cdefs.h>
#ifndef lint
#if 0
static char rcsid[] =
    "@(#) Header: mrinfo.c,v 1.6 93/04/08 15:14:16 van Exp (LBL)";
#else
__RCSID("$NetBSD: mrinfo.c,v 1.26 2007/02/22 01:29:35 hubertf Exp $");
#endif
#endif

#include <ctype.h>
#include <string.h>
#include <netdb.h>
#include <sys/time.h>
#include <poll.h>
#include "defs.h"
#include <arpa/inet.h>
#include <stdarg.h>

#define DEFAULT_TIMEOUT	4	/* How long to wait before retrying requests */
#define DEFAULT_RETRIES 3	/* How many times to ask each router */

u_int32_t	our_addr, target_addr = 0;	/* in NET order */
int     debug = 0;
int	nflag = 0;
int     retries = DEFAULT_RETRIES;
int     timeout = DEFAULT_TIMEOUT;
int	target_level = 0;
vifi_t  numvifs;		/* to keep loader happy */
				/* (see COPY_TABLES macro called in kern.c) */

char *			inet_name(u_int32_t addr);
void			ask(u_int32_t dst);
void			ask2(u_int32_t dst);
int			get_number(int *var, int deflt, char ***pargv,
				   int *pargc);
u_int32_t		host_addr(char *name);
void			usage(void);

/* to shut up -Wstrict-prototypes */
int			main(int argc, char *argv[]);
/* logit() prototyped in defs.h */


char   *
inet_name(u_int32_t addr)
{
	struct hostent *e;
	struct in_addr in;

	if (addr == 0)
		return "local";

	if (nflag ||
	    (e = gethostbyaddr((char *)&addr, sizeof(addr), AF_INET)) == NULL) {
		in.s_addr = addr;
		return (inet_ntoa(in));
	}
	return (e->h_name);
}

/*
 * Log errors and other messages to stderr, according to the severity of the
 * message and the current debug level.  For errors of severity LOG_ERR or
 * worse, terminate the program.
 */
void
logit(int severity, int syserr, const char *format, ...)
{
	va_list ap;

	switch (debug) {
	case 0:
		if (severity > LOG_WARNING)
			return;
	case 1:
		if (severity > LOG_NOTICE)
			return;
	case 2:
		if (severity > LOG_INFO)
			return;
	default:
		if (severity == LOG_WARNING)
			fprintf(stderr, "warning - ");
		va_start(ap, format);
		vfprintf(stderr, format, ap);
		va_end(ap);
		if (syserr == 0)
			fprintf(stderr, "\n");
		else
			fprintf(stderr, ": %s\n", strerror(syserr));
	}

	if (severity <= LOG_ERR)
		exit(1);
}

/*
 * Send a neighbors-list request.
 */
void
ask(u_int32_t dst)
{
	send_igmp(our_addr, dst, IGMP_DVMRP, DVMRP_ASK_NEIGHBORS,
			htonl(MROUTED_LEVEL), 0);
}

void
ask2(u_int32_t dst)
{
	send_igmp(our_addr, dst, IGMP_DVMRP, DVMRP_ASK_NEIGHBORS2,
			htonl(MROUTED_LEVEL), 0);
}

/*
 * Process an incoming neighbor-list message.
 */
void
accept_neighbors(u_int32_t src, u_int32_t dst, u_char *p, int datalen,
		 u_int32_t level)
{
	u_char *ep = p + datalen;
#define GET_ADDR(a) (a = ((u_int32_t)*p++ << 24), a += ((u_int32_t)*p++ << 16),\
		     a += ((u_int32_t)*p++ << 8), a += *p++)

	printf("%s (%s):\n", inet_fmt(src), inet_name(src));
	while (p < ep) {
		u_int32_t laddr;
		u_char metric;
		u_char thresh;
		int ncount;

		GET_ADDR(laddr);
		laddr = htonl(laddr);
		metric = *p++;
		thresh = *p++;
		ncount = *p++;
		while (--ncount >= 0) {
			u_int32_t neighbor;
			GET_ADDR(neighbor);
			neighbor = htonl(neighbor);
			printf("  %s -> ", inet_fmt(laddr));
			printf("%s (%s) [%d/%d]\n",
				inet_fmt(neighbor),
			       inet_name(neighbor), metric, thresh);
		}
	}
}

void
accept_neighbors2(u_int32_t src, u_int32_t dst, u_char *p, int datalen,
		  u_int32_t level)
{
	u_char *ep = p + datalen;
	u_int broken_cisco = ((level & 0xffff) == 0x020a); /* 10.2 */
	/* well, only possibly_broken_cisco, but that's too long to type. */

	printf("%s (%s) [version %d.%d", inet_fmt(src),
		inet_name(src), level & 0xff, (level >> 8) & 0xff);
	if ((level >> 16) & NF_LEAF)   { printf (",leaf"); }
	if ((level >> 16) & NF_PRUNE)  { printf (",prune"); }
	if ((level >> 16) & NF_GENID)  { printf (",genid"); }
	if ((level >> 16) & NF_MTRACE) { printf (",mtrace"); }
	printf ("]:\n");

	while (p < ep) {
		u_char metric;
		u_char thresh;
		u_char flags;
		int ncount;
		u_int32_t laddr = *(u_int32_t*)p;

		p += 4;
		metric = *p++;
		thresh = *p++;
		flags = *p++;
		ncount = *p++;
		if (broken_cisco && ncount == 0)	/* dumb Ciscos */
			ncount = 1;
		if (broken_cisco && ncount > 15)	/* dumb Ciscos */
			ncount = ncount & 0xf;
		while (--ncount >= 0 && p < ep) {
			u_int32_t neighbor = *(u_int32_t*)p;
			p += 4;
			printf("  %s -> ", inet_fmt(laddr));
			printf("%s (%s) [%d/%d",
				inet_fmt(neighbor),
				inet_name(neighbor), metric, thresh);
			if (flags & DVMRP_NF_TUNNEL)
				printf("/tunnel");
			if (flags & DVMRP_NF_SRCRT)
				printf("/srcrt");
			if (flags & DVMRP_NF_PIM)
				printf("/pim");
			if (flags & DVMRP_NF_QUERIER)
				printf("/querier");
			if (flags & DVMRP_NF_DISABLED)
				printf("/disabled");
			if (flags & DVMRP_NF_DOWN)
				printf("/down");
			if (flags & DVMRP_NF_LEAF)
				printf("/leaf");
			printf("]\n");
		}
	}
}

int
get_number(int *var, int deflt, char ***pargv, int *pargc)
{
	if ((*pargv)[0][2] == '\0') {	/* Get the value from the next
					 * argument */
		if (*pargc > 1 && isdigit((unsigned char)(*pargv)[1][0])) {
			(*pargv)++, (*pargc)--;
			*var = atoi((*pargv)[0]);
			return 1;
		} else if (deflt >= 0) {
			*var = deflt;
			return 1;
		} else
			return 0;
	} else {		/* Get value from the rest of this argument */
		if (isdigit((unsigned char)(*pargv)[0][2])) {
			*var = atoi((*pargv)[0] + 2);
			return 1;
		} else {
			return 0;
		}
	}
}

void
usage(void)
{
	fprintf(stderr,
	    "usage: mrinfo [-n] [-t timeout] [-r retries] [router]\n");
	exit(1);
}

int
main(int argc, char *argv[])
{
	int tries;
	int trynew;
	struct timeval et;
	struct hostent *hp;
	struct hostent bogus;
	char *host;
	int curaddr;

	if (geteuid() != 0) {
		fprintf(stderr, "mrinfo: must be root\n");
		exit(1);
	}
	init_igmp();
	if (setuid(getuid()) == -1)
		logit(LOG_ERR, errno, "setuid");

	setlinebuf(stderr);

	argv++, argc--;
	while (argc > 0 && argv[0][0] == '-') {
		switch (argv[0][1]) {
		case 'd':
			if (!get_number(&debug, DEFAULT_DEBUG, &argv, &argc))
				usage();
			break;
		case 'n':
			++nflag;
			break;
		case 'r':
			if (!get_number(&retries, -1, &argv, &argc))
				usage();
			break;
		case 't':
			if (!get_number(&timeout, -1, &argv, &argc))
				usage();
			break;
		default:
			usage();
		}
		argv++, argc--;
	}
	if (argc > 1)
		usage();
	if (argc == 1)
		host = argv[0];
	else
		host = "127.0.0.1";

	if ((target_addr = inet_addr(host)) != -1) {
		hp = &bogus;
		hp->h_length = sizeof(target_addr);
		hp->h_addr_list = (char **)malloc(2 * sizeof(char *));
		if (hp->h_addr_list == NULL)
			logit(LOG_ERR, errno, "malloc");
		hp->h_addr_list[0] = malloc(hp->h_length);
		if (hp->h_addr_list[0] == NULL)
			logit(LOG_ERR, errno, "malloc");
		memcpy(hp->h_addr_list[0], &target_addr, sizeof(hp->h_addr_list[0]));
		hp->h_addr_list[1] = NULL;
	} else
		hp = gethostbyname(host);

	if (hp == NULL || hp->h_length != sizeof(target_addr)) {
		fprintf(stderr, "mrinfo: %s: no such host\n", argv[0]);
		exit(1);
	}
	if (debug)
		fprintf(stderr, "Debug level %u\n", debug);

	/* Check all addresses; mrouters often have unreachable interfaces */
	for (curaddr = 0; hp->h_addr_list[curaddr] != NULL; curaddr++) {
	    memcpy(&target_addr, hp->h_addr_list[curaddr], sizeof(target_addr));
	    {			/* Find a good local address for us. */
		int     udp;
		struct sockaddr_in addr;
		socklen_t addrlen = sizeof(addr);

		memset(&addr, 0, sizeof(addr));
		addr.sin_family = AF_INET;
#if (defined(BSD) && (BSD >= 199103))
		addr.sin_len = sizeof addr;
#endif
		addr.sin_addr.s_addr = target_addr;
		addr.sin_port = htons(2000);	/* any port over 1024 will
						 * do... */
		if ((udp = socket(AF_INET, SOCK_DGRAM, 0)) < 0
		|| connect(udp, (struct sockaddr *) & addr, sizeof(addr)) < 0
		    || getsockname(udp, (struct sockaddr *) & addr, &addrlen) < 0) {
			perror("Determining local address");
			exit(1);
		}
		close(udp);
		our_addr = addr.sin_addr.s_addr;
	    }

	    tries = 0;
	    trynew = 1;
	    /*
	     * New strategy: send 'ask2' for two timeouts, then fall back
	     * to 'ask', since it's not very likely that we are going to
	     * find someone who only responds to 'ask' these days
	     */
	    ask2(target_addr);

	    gettimeofday(&et, 0);
	    et.tv_sec += timeout;

	    /* Main receive loop */
	    for (;;) {
		struct pollfd set[1];
		struct timeval tv, now;
		int     count, recvlen;
		socklen_t dummy;
		u_int32_t src, dst, group;
		struct ip *ip;
		struct igmp *igmp;
		int     ipdatalen, iphdrlen, igmpdatalen;

		set[0].fd = igmp_socket;
		set[0].events = POLLIN;

		gettimeofday(&now, 0);
		tv.tv_sec = et.tv_sec - now.tv_sec;
		tv.tv_usec = et.tv_usec - now.tv_usec;

		if (tv.tv_usec < 0) {
			tv.tv_usec += 1000000L;
			--tv.tv_sec;
		}
		if (tv.tv_sec < 0)
			tv.tv_sec = tv.tv_usec = 0;

		count = poll(set, 1, tv.tv_sec * 1000 + tv.tv_usec / 1000);

		if (count < 0) {
			if (errno != EINTR)
				perror("select");
			continue;
		} else if (count == 0) {
			logit(LOG_DEBUG, 0, "Timed out receiving neighbor lists");
			if (++tries > retries)
				break;
			/* If we've tried ASK_NEIGHBORS2 twice with
			 * no response, fall back to ASK_NEIGHBORS
			 */
			if (tries == 2 && target_level == 0)
				trynew = 0;
			if (target_level == 0 && trynew == 0)
				ask(target_addr);
			else
				ask2(target_addr);
			gettimeofday(&et, 0);
			et.tv_sec += timeout;
			continue;
		}
		recvlen = recvfrom(igmp_socket, recv_buf, RECV_BUF_SIZE,
				   0, NULL, &dummy);
		if (recvlen <= 0) {
			if (recvlen && errno != EINTR)
				perror("recvfrom");
			continue;
		}

		if (recvlen < sizeof(struct ip)) {
			logit(LOG_WARNING, 0,
			    "packet too short (%u bytes) for IP header",
			    recvlen);
			continue;
		}
		ip = (struct ip *) recv_buf;
		if (ip->ip_p == 0)
			continue;	/* Request to install cache entry */
		src = ip->ip_src.s_addr;
		dst = ip->ip_dst.s_addr;
		iphdrlen = ip->ip_hl << 2;
		ipdatalen = ip->ip_len;
		if (iphdrlen + ipdatalen != recvlen) {
		    logit(LOG_WARNING, 0,
		      "packet shorter (%u bytes) than hdr+data length (%u+%u)",
		      recvlen, iphdrlen, ipdatalen);
		    continue;
		}
		igmp = (struct igmp *) (recv_buf + iphdrlen);
		group = igmp->igmp_group.s_addr;
		igmpdatalen = ipdatalen - IGMP_MINLEN;
		if (igmpdatalen < 0) {
		    logit(LOG_WARNING, 0,
			"IP data field too short (%u bytes) for IGMP, from %s",
			ipdatalen, inet_fmt(src));
		    continue;
		}
		if (igmp->igmp_type != IGMP_DVMRP)
			continue;

		switch (igmp->igmp_code) {
		case DVMRP_NEIGHBORS:
		case DVMRP_NEIGHBORS2:
			if (src != target_addr) {
				fprintf(stderr, "mrinfo: got reply from %s",
					inet_fmt(src));
				fprintf(stderr, " instead of %s\n",
					inet_fmt(target_addr));
				/*continue;*/
			}
			break;
		default:
			continue;	/* ignore all other DVMRP messages */
		}

		switch (igmp->igmp_code) {

		case DVMRP_NEIGHBORS:
			if (group) {
				/* knows about DVMRP_NEIGHBORS2 msg */
				if (target_level == 0) {
					target_level = ntohl(group);
					ask2(target_addr);
				}
			} else {
				accept_neighbors(src, dst, (u_char *)(igmp + 1),
						 igmpdatalen, ntohl(group));
				exit(0);
			}
			break;

		case DVMRP_NEIGHBORS2:
			accept_neighbors2(src, dst, (u_char *)(igmp + 1),
					  igmpdatalen, ntohl(group));
			exit(0);
		}
	    }
	}
	exit(1);
}

/* dummies */
void accept_probe(u_int32_t src, u_int32_t dst, char *p, int datalen,
		  u_int32_t level)
{
}
void accept_group_report(u_int32_t src, u_int32_t dst, u_int32_t group,
			 int r_type)
{
}
void accept_neighbor_request2(u_int32_t src, u_int32_t dst)
{
}
void accept_report(u_int32_t src, u_int32_t dst, char *p, int datalen,
		   u_int32_t level)
{
}
void accept_neighbor_request(u_int32_t src, u_int32_t dst)
{
}
void accept_prune(u_int32_t src, u_int32_t dst, char *p, int datalen)
{
}
void accept_graft(u_int32_t src, u_int32_t dst, char *p, int datalen)
{
}
void accept_g_ack(u_int32_t src, u_int32_t dst, char *p, int datalen)
{
}
void add_table_entry(u_int32_t origin, u_int32_t mcastgrp)
{
}
void check_vif_state(void)
{
}
void accept_leave_message(u_int32_t src, u_int32_t dst, u_int32_t group)
{
}
void accept_mtrace(u_int32_t src, u_int32_t dst, u_int32_t group, char *data,
		   u_int no, int datalen)
{
}
void accept_membership_query(u_int32_t src, u_int32_t dst, u_int32_t group,
			     int tmo)
{
}
void accept_info_request(u_int32_t src, u_int32_t dst, u_char *p, int datalen)
{
}
void accept_info_reply(u_int32_t src, u_int32_t dst, u_char *p, int datalen)
{
}
